src/dataflow/src/server.rs

```diff
@@ -218,19 +217,216 @@ pub fn serve(config: Config) -> Result<WorkerGuards<()>, String> {
                 pending_peeks: Vec::new(),
                 feedback_tx: None,
                 reported_frontiers: HashMap::new(),
+                reported_source_frontiers: HashMap::new(),
                 metrics: Metrics::for_worker_id(worker_idx),
             }
             .run()
         },
     )
 }
 
+/// This struct holds per-source timestamp state in a way that can be shared across
+/// different source instances and allow different source instances to indicate
+/// how far they have read up to.
+///
+/// This type is almost never meant to be used directly, and you probably want to
+/// use `TimestampBindingRc` instead.
+pub struct TimestampBindingBox {
+    /// List of timestamp bindings per independent partition. This vector is sorted
+    /// by timestamp and offset and each `(time, offset)` entry indicates that offsets <=
+    /// `offset` should be assigned `time` as their timestamp. Consecutive entries form
+    /// an interval of offsets.
+    partitions: HashMap<PartitionId, Vec<(Timestamp, MzOffset)>>,
+    /// Indicates the lowest timestamp across all partitions that we retain bindings for.
+    /// This frontier can be held back by other entities holding the shared
+    /// `TimestampBindingRc`.
+    compaction_frontier: MutableAntichain<Timestamp>,
+}
+
+impl TimestampBindingBox {
+    /// Create a new instance of `TimestampBindingBox`.
+    fn new() -> Self {
+        Self {
+            partitions: HashMap::new(),
+            compaction_frontier: MutableAntichain::new(),
+        }
+    }
+
+    fn adjust_compaction_frontier(
+        &mut self,
+        remove: AntichainRef<Timestamp>,
+        add: AntichainRef<Timestamp>,
+    ) {
+        self.compaction_frontier
+            .update_iter(remove.iter().map(|t| (*t, -1)));
+        self.compaction_frontier
+            .update_iter(add.iter().map(|t| (*t, 1)));
+    }
+
+    /// Drop all of the timestamp bindings for timestamps not in advance of the
+    /// compaction frontier.
+    fn compact(&mut self) {
+        let frontier = self.compaction_frontier.frontier();
+
+        if !frontier.is_empty() {
+            // Other folks are using these timestamp bindings. Lets keep everything
+            // >= what people are using
+            for (_, entries) in self.partitions.iter_mut() {
+                entries.retain(|(time, _)| frontier.less_equal(time));
+            }
+        }
+    }
+
+    /// Add a new mapping from `(partition, offset) -> timestamp`.
+    ///
+    /// Note that the `offset` and `timestamp` both have to be greater than the
+    /// largest previously bound offsets or timestamps for that partition.
```

Comment:
It looks like in the code that the offset must be `ge` but the timestamp must be `gt`.

Reply:
updated the comments